### PR TITLE
add absent function

### PR DESCRIPTION
**Pull request by siddhikhapare**
Signed-off-by: siddhikhapare <siddhikhapare66@bvcoenm.edu.in>

Fixes: #138

### CODE REVIEW

**Comment by siddhikhapare** [comment]
@fpetkovski  Can you review it once? If any changes required please let me know.

File: engine/engine_test.go [hunk 1]
```diff
@@ -1500,6 +1500,16 @@ func TestQueriesAgainstOldEngine(t *testing.T) {
 			end:   time.Unix(3000, 0),
 			step:  2 * time.Second,
 		},
+		{
+			name: "absent",
+			load: `load 30s
+				absent{pod="nginx-1", series="1"} 0
+				absent{pod="nginx-2", series="1"} 0`,
+			query: "absent(absent)",
```
Comment (yeya24) [inline engine/engine_test.go:1508]: Can you add another test case where the metric doesn't exist? Also let's rename the metric name to not be absent

File: engine/engine_test.go [hunk 2]
```diff
@@ -2431,6 +2441,15 @@ func TestInstantQuery(t *testing.T) {
 				http_requests_total{pod="nginx-2", series="1"} -10+1x50`,
 			query: "sgn(http_requests_total)",
 		},
+		{
+			name: "absent",
+			load: `load 30s
+			absent{pod="nginx-1", series="1"} 0
+			absent{pod="nginx-2", series="1"} 0`,
+			query: "absent(absent)",
```
Comment (yeya24) [inline engine/engine_test.go:2449]: Ditto

**Comment by siddhikhapare** [comment]
@yeya24  I have made changes as you suggested. Can you review it. Thank you in advanced.

**Comment by yeya24** [comment]
I don't think so. Please have 2 test cases for `absent`. One is calling `absent` when the queried series is present, one is not. I see you only have the case of absent.

**Comment by siddhikhapare** [comment]
@yeya24  I fixed errors. Can you please review it? Thank you.

**Comment by fpetkovski** [comment]
Looks like tests are still failing. You can run `make test` to reproduce the failures.

**Comment by siddhikhapare** [comment]
ok

**Comment by siddhikhapare** [comment]
`
"sort": func(f FunctionArgs) promql.Sample {
	if len(f.Points) == 0 {
		return InvalidSample
	}

	sortedPoints := make([]promql.Point, len(f.Points))
	copy(sortedPoints, f.Points)
	sort.Slice(sortedPoints, func(i, j int) bool {
		return sortedPoints[i].V < sortedPoints[j].V
	})

	sortedValues := make([]float64, len(sortedPoints))
	for i, point := range sortedPoints {
		sortedValues[i] = point.V
	}

	return promql.Sample{
		Metric: f.Labels,
		Point: promql.Point{
			T: f.StepTime,
			V: sortedValues,
		},
	}
},
`
@fpetkovski  @yeya24  In the implementation of sort, sortedPoints is of type []promql.Point and 
sortedValues is of type []float64. In final return statement, the V field of promql.Point is 
assigned a []float64 type which is not allowed as V is of type float64.
To solve this issue, I modified Point struct to have a slice of float64 instead of a 
single float64 value for V but did not work.
I do not understand while testing it provide me error as can't use sortedValues 
(variable of type []float64) as type float64 in struct literal. Can you give me some 
pointers on this?

**Comment by yeya24** [comment]
I prefer to not modify the existing struct but have a separate operator to deal with sort and sort desc.

Let's focus on absent in this pr and fix tests first.

**Comment by siddhikhapare** [comment]
@yeya24  According prometheus query function documentation I have done changes in test cases. absent function is useful for alerting on when no time series exist for a given metric name and label combination.

In TestQueriesAgainstOldEngine -
```
{
    name: "absent with non-present metric",
    load: "",
    query: "absent(dummy_metric)",
    start: time.Unix(0, 0),
    end: time.Unix(3000, 0),
    step: 10 * time.Second,
},
{
    name: "absent with non-present label",
    load: `load 30s
        http_requests_total{pod="nginx-1", series="1"} 1+1.1x40
        http_requests_total{pod="nginx-2", series="1"} -10+1x50`,
    query: "absent(http_requests_total{pod=\"nginx-3\"})",
    start: time.Unix(0, 0),
    end: time.Unix(3000, 0),
    step: 10 * time.Second,
},
{
    name: "absent with present metric and label",
    load: `load 30s
        http_requests_total{pod="nginx-1", series="1"} 1+1.1x40
        http_requests_total{pod="nginx-2", series="1"} -10+1x50`,
    query: "absent(http_requests_total{pod=\"nginx-1\"})",
    start: time.Unix(0, 0),
    end: time.Unix(3000, 0),
    step: 10 * time.Second,
},

```

In TestInstantQuery -

```
{
    name: "absent with non-present metric",
    load: "",
    query: "absent(dummy_metric)",
    queryTime: time.Unix(50, 0),
    sortByLabels: true,
},
{
    name: "absent with non-present label",
    load: `load 30s
        http_requests_total{pod="nginx-1", series="1"} 1+1.1x40
        http_requests_total{pod="nginx-2", series="1"} -10+1x50`,
    query: "absent(http_requests_total{pod=\"nginx-3\"})",
    queryTime: time.Unix(50, 0),
    sortByLabels: true,
},
{
    name: "absent with present metric and label",
    load: `load 30s
        http_requests_total{pod="nginx-1", series="1"} 1+1.1x40
        http_requests_total{pod="nginx-2", series="1"} -10+1x50`,
    query: "absent(http_requests_total{pod=\"nginx-1\"})",
    queryTime: time.Unix(50, 0),
    sortByLabels: true,
},

```

But while running test found errors which indicate that the expected and actual results are not matching for the tests. Can you help  me to understand  what are my mistakes to fixed errors. I mentioned my errors are -


RUN TestInstantQuery/disableOptimizers=false/absent_with_non-
present_metric#01/disableFallback=false

testutil.go:91: engine_test.go:2586: ""
exp: &promql.Result{Err:error(nil),
Value:promql.Vector{promql.Sample{Point:promql.Point{T:50000, V:1, H:(*histogram.FloatHistogram)
(nil)}, Metric:labels.Labels{}}}, Warnings:storage.Warnings(nil)}
got: &promql.Result{Err:error(nil), Value:promql.Vector{}, Warnings:storage.Warnings(nil)}
Diff:
--- Expected
+++ Actual
@@ -1,2 +1,2 @@
-(*promql.Result)({} => 1 @[50000])
+(*promql.Result)()


=== RUN TestInstantQuery/disableOptimizers=false/absent_with_non-
present_label#01/disableFallback=true

testutil.go:91: engine_test.go:2586: ""
exp: &promql.Result{Err:error(nil),
Value:promql.Vector{promql.Sample{Point:promql.Point{T:50000, V:1, H:(*histogram.FloatHistogram)
(nil)}, Metric:labels.Labels{labels.Label{Name:"pod", Value:"nginx-3"}}}},
Warnings:storage.Warnings(nil)}
got: &promql.Result{Err:error(nil), Value:promql.Vector{}, Warnings:storage.Warnings(nil)}
Diff:
--- Expected
+++ Actual
@@ -1,2 +1,2 @@
-(*promql.Result)({pod="nginx-3"} => 1 @[50000])
+(*promql.Result)()


=== RUN TestInstantQuery/disableOptimizers=false/absent_with_present_metric_and_label#01
=== RUN
TestInstantQuery/disableOptimizers=false/absent_with_present_metric_and_label#01/disableFallback=false
testutil.go:91: engine_test.go:2586: ""
exp: &promql.Result{Err:error(nil), Value:promql.Vector{}, Warnings:storage.Warnings(nil)}
got: &promql.Result{Err:error(nil),
Value:promql.Vector{promql.Sample{Point:promql.Point{T:50000, V:0, H:(*histogram.FloatHistogram)
(nil)}, Metric:labels.Labels{labels.Label{Name:"pod", Value:"nginx-1"}, labels.Label{Name:"series",
Value:"1"}}}}, Warnings:storage.Warnings(nil)}
Diff:
--- Expected
+++ Actual
@@ -1,2 +1,2 @@
-(*promql.Result)()
+(*promql.Result)({pod="nginx-1", series="1"} => 0 @[50000])

**Comment by siddhikhapare** [comment]
@yeya24  I am new to this project. I am exploring it. Thank you.

**Comment by yeya24** [comment]
@siddhikhapare The test error is clear that your test result is wrong. The expected value 1 is not returned when it should be.
Using the debugger to see how the engine is implemented could help you understand how to fix it.

File: engine/engine_test.go [hunk 3]
```diff
@@ -1500,6 +1500,24 @@ func TestQueriesAgainstOldEngine(t *testing.T) {
 			end:   time.Unix(3000, 0),
 			step:  2 * time.Second,
 		},
+		{
+			name: "absent with present series",
+			load: `load 30s
+				absent{pod="nginx-1", series="1"} 0
+				absent{pod="nginx-2", series="1"} 0`,
+			query: "absent(metric)",
```
Comment (yeya24) [inline engine/engine_test.go:1508]: This test case seems wrong itself. Your metric name is `absent` but you are calling `absent` for `metric`. `metric` is not present in your test data.

File: engine/engine_test.go [hunk 4]
```diff
@@ -2431,6 +2449,22 @@ func TestInstantQuery(t *testing.T) {
 				http_requests_total{pod="nginx-2", series="1"} -10+1x50`,
 			query: "sgn(http_requests_total)",
 		},
+		{
+			name: "absent with present series",
+			load: `load 30s
+			absent{pod="nginx-1", series="1"} 0
+			absent{pod="nginx-2", series="1"} 0`,
+			query: "absent(metric)",
```
Comment (yeya24) [inline engine/engine_test.go:2457]: ditto

File: execution/function/functions.go [hunk 5]
```diff
@@ -434,6 +434,22 @@ var Funcs = map[string]FunctionCall{
...
+					T: f.StepTime,
+					V: 1,
+				},
+			}
+		}
+
+		return promql.Sample{
+			Metric: f.Labels,
+			Point: promql.Point{},
+		}
```
Comment (yeya24) [inline execution/function/functions.go:451]: I don't think this is the correct. `absent` checks whether the series exists ot not, not samples. When the series exists, I think what you want is to drop the whole series, not returning an empty point. Implementing a function like this couldn't achieve the goal of dropping the series right now, you have to refactor the framework to do that.

Probably a dedicated `absent` operator is easier.

Comment (siddhikhapare) [reply execution/function/functions.go:451]: @yeya24 Thank you for the guidance. I will go through @MichaHoffmann implementation to understand it.

**Pull request closed without merging**